File: app/src/main/java/org/szymonbultrowicz/olympusphototransfer/lib/client/PhotoInfo.kt
package org.szymonbultrowicz.olympusphototransfer.lib.client

import java.net.URL

data class PhotoInfo(
    val files: Set<FileInfo>
): BasePhotoInfo {
    init {
        require(files.isNotEmpty()) {
            "PhotoInfo files set must not be empty"
        }
    }

    override val name = files.first().baseFileName
    override val dateTaken = files.first().dateTaken
    override val thumbnailUrl: URL? = files.firstOrNull { it.thumbnailUrl !== null }?.thumbnailUrl

    val hasRaw = files.any { f -> f.extension == ORF_EXT }
    val hasJpg = files.any { f -> f.extension == JPG_EXT }

    companion object {
        const val ORF_EXT = "orf"
        const val JPG_EXT = "jpg"
    }
}

File: src/lib/camera-connector/photo-info.ts
import { BasePhotoInfo } from './base-photo-info';
import { PhotoFileInfo } from './photo-file-info';
import { isEmpty } from 'lodash-es';

export class PhotoInfo implements BasePhotoInfo {
    constructor(readonly files: PhotoFileInfo[]) {
        if (isEmpty(files)) {
            throw new Error('Photo files list cannot be empty');
        }
    }

    private get firstFile(): PhotoFileInfo {
        return this.files[0];
    }

    get name() {
        return this.firstFile.baseFileName;
    }

    get dateTaken() {
        return this.firstFile.dateTaken;
    }

    get thumbnailUrl() {
        return this.firstFile.thumbnailUrl;
    }
}
